### PR TITLE
(feat) O3-3061: Support for conditional answered validation

### PR DESCRIPTION
**Pull request by arodidev**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [x] My work includes tests or is validated by existing tests.

## Summary

This PR aims to introduce conditional answered validation to the form's engine.

## Screenshots

https://github.com/openmrs/openmrs-form-engine-lib/assets/51090527/ab086194-db2c-40f3-88ec-dc96e29f7b7c

## Related Issue

https://openmrs.atlassian.net/browse/O3-3061

### CODE REVIEW

**Comment by arodidev** [comment]
Opened this work for review purposes pending tests.

File: src/hooks/useFieldValidationResults.ts [hunk 1]
```diff
@@ -6,9 +6,11 @@ export function useFieldValidationResults(field: FormField) {
   const [warnings, setWarnings] = useState([]);
 
   useEffect(() => {
-    if (field.meta?.submission) {
-      setErrors((prevErrors) => [...prevErrors, ...(field.meta.submission.errors || [])]);
-      setWarnings((prevWarnings) => [...prevWarnings, ...(field.meta.submission.warnings || [])]);
+    if (field.meta?.submission?.errors) {
```
Comment (samuelmale) [inline src/hooks/useFieldValidationResults.ts:9]: The meta object is initialised at a earlier stage so it's not necessary to optionally chain it.

File: src/utils/form-helper.ts [hunk 2]
```diff
@@ -37,6 +37,16 @@ export function isInlineView(
   return renderingType == 'single-line';
 }
 
+export function evaluateConditionalAnswered(field: FormField, flattenedFields: FormField[]) {
```
Comment (samuelmale) [inline src/utils/form-helper.ts:40]: `allFields` reads better

File: src/validators/conditional-answered-validator.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { type FormFieldValidator, type FormField } from '../types';
+
+export const conditionalAnsweredValidator: FormFieldValidator = {
+  validate: function (field: FormField, value: unknown, config: Record<string, any>) {
+    const { referenceQuestionId, referenceQuestionAnswers, values, fields, message } = config;
+
+    const referencedField = fields.find((field) => field.id === referenceQuestionId);
+    const referencedFieldValue = values[referencedField.id] || referencedField.meta?.submission?.newValue?.value;
```
Comment (samuelmale) [inline src/validators/conditional-answered-validator.ts:8]: The second part of the expressions presupposes the underlying field to be capturing an obs. I think referencing the `values` object should suffice.

File: src/validators/conditional-answered-validator.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+    if (!referencedFieldValue || !referenceQuestionAnswers.includes(referencedFieldValue)) {
```
Comment (samuelmale) [inline src/validators/conditional-answered-validator.ts:10]: Shouldn't we being doing something like:

```ts
if (!isEmpty(value) && !referenceQuestionAnswers.includes(referencedFieldValue)) {
    return [{ resultType: 'error', errCode: 'invalid.valueSelected', message: message }];
}
```

File: src/form-engine.test.tsx [hunk 5]
```diff
@@ -156,6 +157,52 @@ describe('Form engine component', () => {
     });
   });
 
+  describe('conditional answered validation', () => {
+    it('should fail if target field has a value but no value is selected on the referenced field', () => {});
```
Comment (CynthiaKamau) [inline src/form-engine.test.tsx:161]: Do you need this?

**Review by CynthiaKamau — COMMENTED** [review]
Thanks @arodidev , left a comment

File: src/form-engine.test.tsx (hunk 5 above)
Comment (arodidev) [reply src/form-engine.test.tsx]: Done @CynthiaKamau

**Review by CynthiaKamau — APPROVED** [review]

**Pull request merged**